### PR TITLE
Superoptimizer Blog Post for Victor and Jonathan

### PR DESCRIPTION
**Pull request by JonathanDLTran**
This is the blog post for the _Superoptimizer: a look at the smallest program_ paper for Victor Giannakouris and Jonathan Tran.

### CODE REVIEW

File: content/blog/2022-04-26-Superoptimizer.md [hunk 1]
```diff
@@ -0,0 +1,83 @@
++++
+title = "Superoptimizer: a look at the smallest program"
```
Comment (sampsyo) [inline content/blog/2022-04-26-Superoptimizer.md]: We usually use Title Case for these blog posts.
```suggestion
title = "Superoptimizer: A Look at the Smallest Program"
```

File: content/blog/2022-04-26-Superoptimizer.md [hunk 2]
```diff
@@ -0,0 +1,83 @@
...
+[[extra.authors]]
+name = "Jonathan Tran"
+[[extra.authors]]
+name = "Victor Giannakouris"
++++
+
+## Introduction
+
+The majority of compilers utilize optimization to improve code, in terms of runtime performance, code size, and other factors. However, optimization does not necessarily generate the optimal program; the optimization process merely generates a better candidate program. In the Superoptimization paper by Massalin,, a different paradigm is introduced: optimization is performed to generate the minimal sized program, producing programs that are smaller compared to traditionally compiler-optimizer programs, as well as human hand-optimized programs [^1].
```
Comment (sampsyo) [inline content/blog/2022-04-26-Superoptimizer.md]: Footnotes go after the period. https://capra.cs.cornell.edu/styleguide/#footnote (Applies several places below too.)

Comment (sampsyo) [inline content/blog/2022-04-26-Superoptimizer.md]: utilize -> use https://capra.cs.cornell.edu/styleguide/#simplify

File: content/blog/2022-04-26-Superoptimizer.md [hunk 3]
```diff
@@ -0,0 +1,83 @@
...
+
+As an illustration of Superoptimizer’s capabilities, consider an example centered on compiling the signum function, which is defined as follows: 
+
+```int signum(x) {
```
Comment (sampsyo) [inline content/blog/2022-04-26-Superoptimizer.md]: I think you need a line break after the code fence to make this render correctly.

File: content/blog/2022-04-26-Superoptimizer.md [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+	If (x > 0) return 1;
+	Else if (x <  0) return -1;
+	Else return 0;
```
Comment (sampsyo) [inline content/blog/2022-04-26-Superoptimizer.md]: I know this is pseudocode, but I can't help but notice it would be valid C if the keywords were not capitalized (If -> if, etc.)…

File: content/blog/2022-04-26-Superoptimizer.md [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+}
+```
+
+The author notes this program usually compiles to around 9 instructions, including several conditional jumps, while a human can manually optimize this program to 6 instructions. Remarkably, Superoptimizer finds an even smaller solution, consisting of 4 instructions, which is shown below.
+
+```
+Add.l d0, d0
+Subx.l d1, d1
+Negx.l d0
+Addx.l d1, d1
```
Comment (sampsyo) [inline content/blog/2022-04-26-Superoptimizer.md]: TBH it also seems slightly weird to see capitalization in assembly listings, which are almost always all-lowercase.

File: content/blog/2022-04-26-Superoptimizer.md [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+
+## Core Ideas
+
+The core idea of the Superoptimizer paper is to use brute-force enumeration to find the smallest-sized program `P’` that is semantically equivalent to the original program `P`. The enumeration process begins by considering all programs with 1 instruction, and checks if the program is equivalent to the original program. If none of the programs of size 1 are equivalent to the original program, then programs composed of 2 instructions are considered. This process continues until an enumerated program is found to be semantically equivalent to the original program. By following this procedure, it is guaranteed that the discovered program will be the smallest program in size with the same behavior as the original program.
```
Comment (sampsyo) [inline content/blog/2022-04-26-Superoptimizer.md]: If you want, you can set `latex = true` in the configuration at the top and use math for these things, like $P'$ and $P$.

File: content/blog/2022-04-26-Superoptimizer.md [hunk 7]
```diff
@@ -0,0 +1,83 @@
...
+
+Massalin uses a clever technique to solve this problem. Since enumerated programs are likely to differ on many inputs from the original program, to eliminate semantically different programs, it suffices to create small test vectors as a filtering mechanism. Test vectors are manually chosen, and Massalin used test vectors that included edge cases for the type of program that was being tested, as well as more-general inputs ranging from -1024 to 1024. 
+
+While test vectors are likely to eliminate many semantically different enumerated programs, some enumerated programs may still fall through the cracks. At this point, it is the responsibility of the operator of the Superoptimizer to manually ensure that the proposed program is equivalent to the original program.
```
Comment (sampsyo) [inline content/blog/2022-04-26-Superoptimizer.md:51]: If you want, this would be a good place to insert your own commentary about how satisfying (or not) the idea of manual checking is (and how subsequent superoptimizers have dealt with the scalability of equivalence checking).

File: content/blog/2022-04-26-Superoptimizer.md [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+
+The Superoptimizer can also be applied to larger functions, with the caveat that some clever work has to be done to allow the Superoptimizer to run in a feasible time frame. For instance, the author applied the Superoptimizer to optimize a `printf` function.  By applying the Superoptimizer to small chunks of the `printf` function, the function could be improved incrementally. In fact, using the prior approach seems even better when one has an idea of what types of code are more frequently used. Having a profiler guide where these small chunks of code can be superoptimized seems like one approach to use this technique. 
+
+Another way the Superoptimizer can be used is to identify sequences of instructions that can be used for pruning. In this use case, the Superoptimizer can enumerate sets of instruction sequences equivalent to a single instruction. When a sequence of instructions is found to be in that set, we know that it is redundant to the original single instruction, and thus can be pruned.
```
Comment (sampsyo) [inline content/blog/2022-04-26-Superoptimizer.md]: Not just equivalent to a single instruction, I think, but equivalent to _any shorter sequence of instructions_, because the definition of "optimality" here is always the shortest instruction sequence.

Comment (sampsyo) [inline content/blog/2022-04-26-Superoptimizer.md]: Delete comma in "instruction, and" (it joins a dependent clause).

File: content/blog/2022-04-26-Superoptimizer.md [hunk 9]
```diff
@@ -0,0 +1,83 @@
...
+
+Finally, the author also gives an instance of using the Superoptimizer to discover equivalent statements in computer architecture. As some instruction set architectures contain many equivalent instructions, it is possible to use the Superoptimizer to search for an equivalent sequence of simpler instructions that are equivalent to a more complex instruction in the ISA. 
+
+## Related Works
+
+The Superoptimizer has several limitations which influence how the Superoptimizer can be applied. One of the major limitations of Massalin’s work is that Superoptimizer takes a long time to run and find optimal programs. Another downside is that Superoptimizer only works on the Motorola 68000 and Intel 8086 architectures. Finally, yet another issue with Superoptimizer is that enumerated programs are not guaranteed to be equivalent to the original program. These downsides influence the design of future projects that build upon the original Superoptimizer.
```
Comment (sampsyo) [inline content/blog/2022-04-26-Superoptimizer.md]: While it is not wrong to say that scalability (running for a very long time on all but the smallest programs) is a limitation "of Massalin's work," it would be even more correct/helpful to say this is _a glaring limitation of the entire concept_ that should be 100% obvious to anyone with a CS background who hears about the idea for the first time. It's what makes this such a wild idea!! So it's not like this is an unfortunate flaw in the current proposal; it is a fundamental trade-off that can absolutely never be fixed completely. So what's surprising is that it works at all _despite_ this obvious problem.

Comment (sampsyo) [inline content/blog/2022-04-26-Superoptimizer.md]: Also, I'm not 100% sure we should count the target architecture support as a limitation… yes, technically it is limited in the sense that there are other architectures that exist in the world, but supporting _all_ architectures would be impossible, right? Maybe a different way to put it would be to just say that she focused on a couple of ISAs, and subsequent work has done the obvious thing of increasing target support.

File: content/blog/2022-04-26-Superoptimizer.md [hunk 10]
```diff
@@ -0,0 +1,83 @@
...
+
+For instance, the GNU Superoptimizer project made the Superoptimizer project portable across many different architectures [^2]. The GNU Superoptimizer also made superoptimization to the commonly used GCC compiler, allowing this concept to be applied to real world programs. This work ultimately expands the scope of the Superoptimizer across many architectures, but it still suffers from the same limitations of the original Superoptimizer. In particular, this project still performs poorly in terms of time to find a superoptimized solution. Even a 7 sequence of instructions took a “few weeks” to generate, according to the authors.
```
Comment (sampsyo) [inline content/blog/2022-04-26-Superoptimizer.md]: Some word in the phrase "made superoptimization to the commonly used GCC compiler" seems incorrect… not entirely sure what the intent was here.

Comment (sampsyo) [inline content/blog/2022-04-26-Superoptimizer.md]: Maybe also worth noting that, despite being a GNU project like GCC, I don't think anyone uses it (at least compared to the number of people who use GCC).

File: content/blog/2022-04-26-Superoptimizer.md [hunk 11]
```diff
@@ -0,0 +1,83 @@
...
+
+Another work that builds upon the Superoptimizer is Denali [^3]. Denali changes the way the search space of programs is discovered. Denali only considers a search space of semantically equivalent programs to the original program. In this sense, the search space is smaller than that considered by the original Superoptimizer, because all programs that differ semantically from the original program are omitted. Denali is able to consider this search space by applying rewrite rules to the original program to generate a large search space. These rewrite rules include rules like `2 * x = x + x`. Rewrite rules, which can be either system axioms or user given axioms, are incrementally applied to the program, to create new programs. These new programs are incrementally stored in a data structure, the E Graph, which is able to represent every possible rewrite in a polynomial amount of space. Finally, Denali is able to query the E Graph for programs satisfying certain conditions, which represent optimal programs.
```
Comment (sampsyo) [inline content/blog/2022-04-26-Superoptimizer.md]: Maybe expand "E Graph" as "equality graph" the first time it's introduced?

File: content/blog/2022-04-26-Superoptimizer.md [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+
+Finally, to improve upon the speed of the Superoptimizer, the STOKE project uses a different search method [^4]. In this project, stochastic search is employed to first search the program space for enumerated programs, and then an optimized program is chosen from these subspaces. This approach is able to scale up to larger programs, using a shorter amount of time to compile, compared to the original Superoptimizer. The authors of STOKE even find that on certain benchmarks, optimized output programs can surpass -O3 and hand optimized programs. Once again, there is a tradeoff where the most minimal program might not be found, as compared to Superoptimizer. However, this tradeoff may not be that important, because STOKE’s ability to beat hand optimized programs, particularly on larger sized programs, allows the ability to bring superoptimization to a larger space of applications, which can benefit more programmers across the spectrum of programming.
```
Comment (sampsyo) [inline content/blog/2022-04-26-Superoptimizer.md]: Probably typeset `-O3` as code?

File: content/blog/2022-04-26-Superoptimizer.md [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+
+## Conclusion
+
+Massalin’s work is seminal, because it is the first to consider truly optimal programs, rather than incremental improvements that constitute most compiler optimizations. Being able to find optimal real world programs such as the signum function was a major success of Massalin’s work. However, Superoptimizer has several downsides such as the time spent searching for programs, and the inability to guarantee superoptimized programs correct by construction. These downsides are dealt with in other works that build upon ideas found in Superoptimizer. All together, the Superoptimizer provided a practical approach, yet novel, to solve an important problem. Today, the techniques pioneered in this paper continue to be used and improved upon to provide new solutions.
```
Comment (sampsyo) [inline content/blog/2022-04-26-Superoptimizer.md]: This is a fine conclusion, but it could be enhanced even further by playing up the fact that the idea seems impractical on its face. A main takeaway in the paper is that, despite this "obvious" impracticality in general, it _can actually work_ in some situations. And that's what has been so inspiring to subsequent work.

**Review by sampsyo — CHANGES_REQUESTED** [review]
Very nice work, y'all! I enjoyed the post, and this did a good job of explaining the main ideas that go into the system as described in the paper. I have a couple of comments in here about adding emphasis on what I think is a main theme: the "obvious" impracticality of the idea, and how clever tricks can beat back this impracticality in restricted situations. See if any of those make sense!

I also made a few granular/low-level comments throughout. Once you've had a chance to address those, please "request a review" from the GitHub PR page to let me know it's ready to merge.

File: content/blog/2022-04-26-Superoptimizer.md (hunk 1 above)
Comment (JonathanDLTran) [reply content/blog/2022-04-26-Superoptimizer.md]: Makes sense! I have changed this.

File: content/blog/2022-04-26-Superoptimizer.md (hunk 2 above)
Comment (JonathanDLTran) [reply content/blog/2022-04-26-Superoptimizer.md]: Fixed!

File: content/blog/2022-04-26-Superoptimizer.md (hunk 4 above)
Comment (JonathanDLTran) [reply content/blog/2022-04-26-Superoptimizer.md]: I've changed this to be C code in the markdown file.

File: content/blog/2022-04-26-Superoptimizer.md (hunk 6 above)
Comment (JonathanDLTran) [reply content/blog/2022-04-26-Superoptimizer.md]: I added the latex=true at the top and changed a few of the mathematical symbols.

File: content/blog/2022-04-26-Superoptimizer.md (hunk 9 above)
Comment (JonathanDLTran) [reply content/blog/2022-04-26-Superoptimizer.md]: I rewrote this not to list the limitations, but to focus on different ideas that build upon the Superoptimizer's core ideas.

File: content/blog/2022-04-26-Superoptimizer.md (hunk 13 above)
Comment (JonathanDLTran) [reply content/blog/2022-04-26-Superoptimizer.md]: I fixed up the conclusion to focus on the point that you made about how the impractical solution can still lead to impressive results when optimization certain programs.

File: content/blog/2022-04-26-Superoptimizer.md (hunk 9 above)
Comment (JonathanDLTran) [reply content/blog/2022-04-26-Superoptimizer.md]: I also removed the target architecture limitation. I agree it does not make sense to support all architectures.

**Comment by JonathanDLTran** [comment]
Thank you for the feedback! I made the changes and there should be a new commit.

**Review by sampsyo — APPROVED** [review]
Nice work! I like the additional commentary about verification.
